### PR TITLE
Initialize the financial market api framework.

### PR DESCRIPTION
**Pull request by songrgg**
Initialize API framework and add financial market APIs.
1. Query financial products API.
2. Create trade for the specified product API

### CODE REVIEW

File: modules/WscnFMarket/config/routes.frontend.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+return [
+    '/' => [
+        'module' => 'WscnFMarket',
```
Comment (chaiyangyang) [inline modules/WscnFMarket/config/routes.frontend.php]: 没用到的path，controller还有action可以下次用到的时候再加吧

File: modules/WscnFMarket/src/WscnFMarket/Controllers/ControllerBase.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: yudoudou
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ControllerBase.php]: User: songrong

File: modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $numberPage = $this->request->getQuery('page', 'int', 1);
+        $perPage = $this->request->getQuery('perPage', 'int', 20);
+        $orderBy = $this->request->getQuery('orderBy', 'string', 'id');
+
+        $arrProducts = ProductModel::multiQuery($numberPage, $perPage, $orderBy);
+
+        return $this->response->setJsonContent(
+            [
+                'code' => 0,
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php]: 把成功代码抽取到ControllerBase里面去，然后赋一个常量

something like:
declare:
ControllerBase:
const RESPONSE_SUCCESS_CODE = 0;

Refer:
ProductController
'code' => ControllerBase::RESPONSE_SUCCESS_CODE;

Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php]: do the same thing to 'message'

File: modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+use Wscn\WscnFMarket\Entities\Products;
+
+class ProductModel
+{
+    public static function getAll()
+    {
+        return Products::find();
+    }
+
+    public static function multiQuery($page=1, $perPage=20, $orderBy='id')
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php]: space:
$page = 1

File: modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+     * @operationName("获取指定页面的金融产品列表")
+     * @operationDescription("获取指定页面的金融产品列表")
+     */
+    public function queryAction()
+    {
+        $numberPage = $this->request->getQuery('page', 'int', 1);
+        $perPage = $this->request->getQuery('perPage', 'int', 20);
+        $orderBy = $this->request->getQuery('orderBy', 'string', 'id');
+
+        $arrProducts = ProductModel::multiQuery($numberPage, $perPage, $orderBy);
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php]: 只用getAll返回，然后不需要任何参数，同时检查返回结果

File: modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Wscn\WscnFMarket\Models;
+
+use Wscn\WscnFMarket\Entities\Products;
+
+class ProductModel
+{
+    public static function getAll()
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php]: rename
getAll -> getAllProducts()

File: modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        foreach ($arrProduct as &$product) {
+            $tags = $product['tags'];
+            if (!empty($tags)) {
+                $tags = json_decode($tags);
+                $product['tags'] = $tags;
+            }
+        }
+
+        return $arrProduct;
+    }
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php]: 不要在返回结果上面修改，创建一个空list并存放修改的结果:

tags用一个class常量来保存
const COLUMN_TAGS = 'tags';

$products = [];
foreach(Products::find($parameters)->toArray() as $product) {
  $tags = $product[COLUMN_TAGS];
  $normalizedProduct = $product;
  if (!empty($tags)) {
   // tags are comma separated items, eg: low risk,high return,T+0
    normalizedProduct[COLUMN_TAGS] = json_decode($tags);
  }
  products.add(normalizedProduct);
}

File: modules/WscnFMarket/config/routes.frontend.php (hunk 1 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/config/routes.frontend.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Controllers/ControllerBase.php (hunk 2 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ControllerBase.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php (hunk 5 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php (hunk 3 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Controllers/ProductsController.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php (hunk 6 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php (hunk 4 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php (hunk 7 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/ProductModel.php]: DONE

File: apps/FMarketApiVer1/Module.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace FMarketApiVer1;
+
+use Phalcon\Mvc\ModuleDefinitionInterface;
+use Eva\EvaEngine\Module\StandardInterface;
```
Comment (chaiyangyang) [inline apps/FMarketApiVer1/Module.php]: use 按照名称排序

File: apps/FMarketApiVer1/src/FMarketApiVer1/Controllers/ProductsController.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+     *       )
+     *     )
+     *   )
+     * )
+     * @operationName("获取指定页面的金融产品列表")
+     * @operationDescription("获取指定页面的金融产品列表")
+     */
+    public function queryAction()
+    {
+        throw new Exception\VerifyFailedException('Access Token user not match');
```
Comment (chaiyangyang) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Controllers/ProductsController.php]: remove debug lines

File: modules/WscnFMarket/src/WscnFMarket/Models/InvestTrade.php [hunk 10]
```diff
@@ -0,0 +1,238 @@
+<?php
+
```
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/InvestTrade.php]: remove this file?

File: apps/FMarketApiVer1/Module.php (hunk 8 above)
Comment (chaiyangyang) [inline apps/FMarketApiVer1/Module.php]: DONE

File: apps/FMarketApiVer1/src/FMarketApiVer1/Controllers/ProductsController.php (hunk 9 above)
Comment (chaiyangyang) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Controllers/ProductsController.php]: DONE

File: modules/WscnFMarket/src/WscnFMarket/Models/InvestTrade.php (hunk 10 above)
Comment (chaiyangyang) [inline modules/WscnFMarket/src/WscnFMarket/Models/InvestTrade.php]: DONE

File: apps/FMarketApiVer1/Module.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    {
+    }
+
+    public function registerServices($di)
+    {
+        // 当此模块加载时，注册本模块controller的命名空间
+        $dispatcher = $di->get('dispatcher');
+        $dispatcher->setDefaultNamespace('FMarketApiVer1\Controllers');
+    }
+}
```
Comment (AlloVince) [inline apps/FMarketApiVer1/Module.php]: 可以直接继承Eva\EvaEngine\Module\AbstractModule

File: modules/WscnFMarket/Module.php [hunk 12]
```diff
@@ -38,6 +38,6 @@ public function registerServices($di)
     {
         // 当此模块加载时，注册本模块controller的命名空间
         $dispatcher = $di->get('dispatcher');
-        $dispatcher->setDefaultNamespace('SolaApi\Controllers');
+        $dispatcher->setDefaultNamespace('Wscn\WscnFMarket\Controllers');
     }
 }
```
Comment (AlloVince) [inline modules/WscnFMarket/Module.php]: 可以直接继承Eva\EvaEngine\Module\AbstractModule

File: apps/FMarketApiVer1/Module.php (hunk 11 above)
Comment (songrgg) [inline apps/FMarketApiVer1/Module.php]: DONE

**Pull request merged**